py/desispec/io/photo.py

```diff
@@ -0,0 +1,862 @@
+                        cachecat = photocache[targetfile][I]
+                    else:
+                        cachecat = Table(fitsio.read(targetfile, rows=I))
+                    
+                    _photo = Table(np.hstack(np.repeat(datamodel, len(I))))
+                    for col in _photo.colnames: # not all these columns will exist...
+                        if col in cachecat.colnames:
+                            _photo[col] = cachecat[col]
+                    photofiles.append(targetfile)
+                    photo.append(_photo)
+                continue
+
+            if 'ToO' in targetfile:
+                photo1 = Table.read(targetfile, guess=False, format='ascii.ecsv')
+                I = np.where(np.isin(photo1['TARGETID'], input_cat['TARGETID']))[0]
+                log.debug('Matched {} TOO targets'.format(len(I)))
+                if len(I) > 0:
+                    photo1 = photo1[I]
+                    _photo = Table(np.hstack(np.repeat(datamodel, len(I))))
+                    for col in _photo.colnames: # not all these columns will exist...
+                        if col in photo1.colnames:
+                            _photo[col] = photo1[col]
+                    del photo1
+                    photofiles.append('TOO')
+                    photo.append(_photo)
+                continue
+
+            # get the correct extension name or number
+            tinfo = fitsio.FITS(targetfile)
+            for _tinfo in tinfo:
+                extname = _tinfo.get_extname()
+                if 'TARGETS' in extname:
+                    break
+            if extname == '':
+                extname = 1
+                
+            # fitsio does not preserve the order of the rows but we'll sort later.
+            photo_targetid = tinfo[extname].read(columns='TARGETID')
+            I = np.where(np.isin(photo_targetid, input_cat['TARGETID']))[0]
+            
+            log.debug('Matched {} targets in {}'.format(len(I), targetfile))
+            if len(I) > 0:
+                photo1 = tinfo[extname].read(rows=I)
+                # Columns can be out of order, so sort them here based on the
+                # data model so we can stack below.
+                _photo = Table(np.hstack(np.repeat(datamodel, len(I))))
+                for col in _photo.colnames: # all these columns should exist...
+                    if col in photo1.dtype.names:
+                        _photo[col] = photo1[col]
+                    else:
+                        log.debug('Skipping missing column {} from {}'.format(col, targetfile))
+                del photo1
+                photofiles.append(targetfile)
+                photo.append(_photo)
+                #if input_cat['TILEID'][0] == 80615:
+                #    pdb.set_trace()
+
+    # backup programs have no target catalog photometry at all
+    if len(photo) == 0:
+        log.warning('No targeting photometry found.')
+        raise ValueError
```

Comment:
Here and elsewhere, normally if a problem is bad enough to raise an exception, it should be logged as critical and not just a warning.  Context:
warning = bad, but expected to sometimes happen
error = bad and not expected to happen, but not fatal either
critical = so bad that you can't proceed

Consider using this pattern so that the error message is both logged and passed back through the traceback:
```
errmsg = 'blah...'
log.critical(errmsg)
raise ValueError(errmsg)
```

Reply:
Done.